[Backport v2.7-branch] Bluetooth: fix HCI command / data deadlock

theob-pro opened the pull request:
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/58412.

The following commits from https://github.com/zephyrproject-rtos/zephyr/pull/50476 have been added:
- https://github.com/zephyrproject-rtos/zephyr/commit/7a6872d83729360d2061957305e68250107c9e64
- https://github.com/zephyrproject-rtos/zephyr/commit/1c8fe67a52766381fa02de4a39b1888c18d63b6a
- https://github.com/zephyrproject-rtos/zephyr/commit/3c1ca93fe8342d004bf539e3c193efb034ba448a
- https://github.com/zephyrproject-rtos/zephyr/commit/8e207fefad2e1aeaaf03ee4da0fb024d167b7a96
- https://github.com/zephyrproject-rtos/zephyr/commit/77e1a9dcadff4555c5fd9d5ecef10cd7d9fb1c28

https://github.com/zephyrproject-rtos/zephyr/pull/52760 is also backported with this PR.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/57947

jori-nordic committed:
Bluetooth: host: add l2cap stress-test

This test reproduces more-or-less #34600.

It has a central that connects to multiple peripherals, opens one l2cap
CoC channel per connection, and transmits a few SDUs largely exceeding
the MPS of the channel.

In this commit, the test doesn't pass, but when it passes (after the
subsequent commits), error and warning messages are expected from the
stack, as this is not the happy path.

We can later debate on whether these particular error messages should
be downgraded to debug.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 7a6872d83729360d2061957305e68250107c9e64)

jori-nordic committed:
Bluetooth: host: l2cap: release segment in all cases

See code comment

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 1c8fe67a52766381fa02de4a39b1888c18d63b6a)

jori-nordic committed:
Bluetooth: host: l2cap: add alloc_seg callback

This callback allows use-cases where the SDU is much larger than the
l2cap MPS. The stack will then try to allocate using this callback if
specified, and fall-back on using the buffer's pool (previous
behavior).

This way one can define two buffer pools, one with a very large buffer
size, and one with a buffer size >= MPS, and the stack will allocate
from that instead.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 77e1a9dcadff4555c5fd9d5ecef10cd7d9fb1c28)

jori-nordic committed:
Bluetooth: host: l2cap: don't send too much credits

There was an edge-case where we were sending back too much credits, add
a check so we can't do that anymore.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 3c1ca93fe8342d004bf539e3c193efb034ba448a)

jori-nordic committed:
Bluetooth: host: l2cap: workaround SDU deadlock

See the code comments.

SDUs might enter a state where they will be blocked forever, as a
workaround, we nudge them when another SDU has been sent.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 8e207fefad2e1aeaaf03ee4da0fb024d167b7a96)

jori-nordic committed:
Bluetooth: host: make HCI fragmentation async

Make the ACL fragmentation asynchronous, freeing the TX thread for
sending commands when the ACL buffers are full.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit c3e5fabbf19850b10012c4450d6fdeff45714f62)

jori-nordic committed:
Bluetooth: host: poll on CTLR buffers instead of host TX queue

When there are no buffers, it doesn't make sense to repeatedly try to
send the host TX queue.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit ef19c64f1b77316ad3bb36ee0921eec4840c8889)

jori-nordic committed:
Bluetooth: host: copy fragment data at the last minute

Only copy the data from the 'parent' buf into the segment buf if we
successfully acquired the HCI semaphore (ie there are available
controller buffers).

This way we avoid pulling and pushing back the data in the case where
there aren't any controller buffers available.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit ca51439cd1ad3864ed781dd78ab5d851d55d0889)

jori-nordic committed:
Bluetooth: host: update l2cap stress test

Do these things:
- use the proper macros for reserving the SDU header
- make every L2CAP PDU fragment into 3 ACL packets
- set tx data and verify rx matches the pattern
- measure segment pool usage

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
(cherry picked from commit 8bc094610c5dd69f6ba0085fb3fe7d18167164c9)

carlescufi committed:
net: buf: Simplify fragment handling

This patch reworks how fragments are handled in the net_buf
infrastructure.

In particular, it removes the union around the node and frags members
in the main net_buf structure. This is done so that both can be used at
the same time, at a cost of 4 bytes per net_buf instance.
This implies that the layout of net_buf instances changes whenever
being inserted into a queue (fifo or lifo) or a linked list (slist).

Until now, this is what happened when enqueueing a net_buf with frags
in a queue or linked list:

1.1 Before enqueueing:

 +--------+      +--------+      +--------+
 |#1  node|\     |#2  node|\     |#3  node|\
 |        | \    |        | \    |        | \
 | frags  |------| frags  |------| frags  |------NULL
 +--------+      +--------+      +--------+

net_buf #1 has 2 fragments, net_bufs #2 and #3. Both the node and frags
pointers (they are the same, since they are unioned) point to the next
fragment.

1.2 After enqueueing:

 +--------+     +--------+     +--------+     +--------+     +--------+
 |q/slist |-----|#1  node|-----|#2  node|-----|#3  node|-----|q/slist |
 |node    |     | *flag  | /   | *flag  | /   |        | /   |node    |
 |        |     | frags  |/    | frags  |/    | frags  |/    |        |
 +--------+     +--------+     +--------+     +--------+     +--------+

When enqueing a net_buf (in this case #1) that contains fragments, the
current net_buf implementation actually enqueues all the fragments (in
this case #2 and #3) as actual queue/slist items, since node and frags
are one and the same in memory. This makes the enqueuing operation
expensive and it makes it impossible to atomically dequeue. The `*flag`
notation here means that the `flags` member has been set to
`NET_BUF_FRAGS` in order to be able to reconstruct the frags pointers
when dequeuing.

After this patch, the layout changes considerably:

2.1 Before enqueueing:

 +--------+       +--------+       +--------+
 |#1  node|--NULL |#2  node|--NULL |#3  node|--NULL
 |        |       |        |       |        |
 | frags  |-------| frags  |-------| frags  |------NULL
 +--------+       +--------+       +--------+

This is very similar to 1.1, except that now node and frags are
different pointers, so node is just set to NULL.

2.2 After enqueueing:

 +--------+      +--------+      +--------+
 |q/slist |------|#1  node|------|q/slist |
 |node    |      |        |      |node    |
 |        |      | frags  |      |        |
 +--------+      +--------+      +--------+
                     |           +--------+       +--------+
                     |           |#2  node|--NULL |#3  node|--NULL
                     |           |        |       |        |
                     +-----------| frags  |-------| frags  |------NULL
                                 +--------+       +--------+

When enqueuing net_buf #1, now we only enqueue that very item, instead
of enqueing the frags as well, since now node and frags are separate
pointers. This simplifies the operation and makes it atomic.

Resolves #52718.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
(cherry picked from commit 3d306c181f0d75c5765f80fdfbee614e1e3f191d)

PavelVPV committed:
bluetooth: mesh: Remove illegal use of NET_BUF_FRAG in friend.c

This commit removes illegal use of NET_BUF_FRAG in friend.c, which is an
internal flag.

Now `struct bt_mesh_friend_seg` keeps pointer to a first received
segment of a segmented message. The rest segments are added as fragments
using net_buf API. Friend Queue keeps only head of the fragments.
When one segment (currently head of fragments) is removed from Friend
Queue, the next segment is added to the queue. Head has always 2
references: one when allocated, another one when added as fragments
head.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
(cherry picked from commit 5d059117fd279ee0eec4fb1c4d0244a37937013c)

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -520,46 +543,61 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
+		 * and not one of its fragments.
+		 * This buffer was fetched from the FIFO using a peek operation.
+		 */
+		buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
```
If I remember, this relied on the patch that removed the `frags` union in `struct net_buf` (https://github.com/zephyrproject-rtos/zephyr/pull/52760).
Maybe backport that too?

jori-nordic reviewed (changes requested):
see comment

theob-pro committed:
Net: Increase `NET_BUF_USER_DATA_SIZE` value to 8

Increase `NET_BUF_USER_DATA_SIZE` value to 8 if `BT_CONN` is enabled.
This is necessary because one of the backported commits adds one struct
member into `struct tx_meta`, and that will be stored in the buffer
user data.

On main, this Kconfig option has been deprecated, hence why this change
was not present in the original patch set.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

cfriedt commented:
@jhedberg - looks to be ready to me, but this will need your approval. Please take a look.

sheindel-llt commented:
Thanks for doing this packport!
